fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/cursors/SizeStatisticsCollectorCursor.java

```diff
@@ -0,0 +1,613 @@
+                }
+            }).whenComplete((ignore, err) -> kvCursor.close()); //no need to keep it open as it must be reconstituted with a continuation later
+        });
+    }
+
+    //update partial stats
+    private void updateStatistics(@Nonnull KeyValue kv) {
+        this.sizeStatisticsResults.keyCount += 1;
+        this.sizeStatisticsResults.keySize += kv.getKey().length;
+        this.sizeStatisticsResults.maxKeySize = Math.max(this.sizeStatisticsResults.maxKeySize, kv.getKey().length);
+        this.sizeStatisticsResults.valueSize += kv.getValue().length;
+        this.sizeStatisticsResults.maxValueSize = Math.max(this.sizeStatisticsResults.maxValueSize, kv.getValue().length);
+        int totalSize = kv.getKey().length + kv.getValue().length;
+        if (totalSize > 0) {
+            this.sizeStatisticsResults.sizeBuckets[Integer.SIZE - Integer.numberOfLeadingZeros(totalSize) - 1] += 1;
+        }
+    }
+
+    @Nonnull
+    @Override
+    public Executor getExecutor() {
+        return context.getExecutor();
+    }
+
+    @Nonnull
+    @Override
+    public boolean accept(@Nonnull RecordCursorVisitor visitor) {
+        visitor.visitEnter(this);
+        return visitor.visitLeave(this);
+    }
+
+    @Nonnull
+    @Override
+    @Deprecated
+    public CompletableFuture<Boolean> onHasNext() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Nullable
+    @Override
+    @Deprecated
+    public SizeStatisticsResults next() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Nullable
+    @Override
+    @Deprecated
+    public byte[] getContinuation() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Nonnull
+    @Override
+    @Deprecated
+    public RecordCursor.NoNextReason getNoNextReason() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void close() {
+        if (hasNextFuture != null) {
+            hasNextFuture.cancel(false);
+        }
+    }
+
+    //form a continuation that allows us to restart statistics aggregation from where we left off
+    private class SizeStatisticsCollectorCursorContinuation implements RecordCursorContinuation {
+        @Nonnull
+        private final RecordCursorResult<KeyValue> currentKvResult;
+        @Nonnull
+        private final Function<byte[], RecordCursorProto.SizeStatisticsContinuation> continuationFunction;
+        @Nonnull
+        private final boolean isEnd = false;
+        @Nonnull
+        private final SizeStatisticsResults sizeStatisticsResults;
+        @Nullable
+        private byte[] cachedBytes;
+
+        private SizeStatisticsCollectorCursorContinuation(RecordCursorResult<KeyValue> currentKvResult, SizeStatisticsResults sizeStatisticsResults) {
+            this.cachedBytes = null;
+            this.currentKvResult = currentKvResult;
+            this.sizeStatisticsResults = sizeStatisticsResults.copy(); //cache an immutable snapshot of the partial aggregate state
```

Comment:
Instead of copying the `sizeStatisticsResults`, would it make sense to instead make the Protobuf message eagerly? Serialization can still be lazy.

I also don't object too much to this approach.

Reply:
ok so you are looking for ... 

public boolean isEnd() {
            return false;

Reply:
that is question :)

Reply:
Yeah, that's what I'm looking for, though it looks like this was added to the wrong comment.

Reply:
yeah I didn't see a box to respond to the previous comment.

Reply:
If the comment is a response to a comment from a previous thread, the most recent comment isn't directly respondable to, but you can respond to it in the first thread: https://github.com/FoundationDB/fdb-record-layer/pull/543#discussion_r277814740

Reply:
Do you have a feel on the original question from this comment thread before it was co-opted by the other issue?

Reply:
I lost track of what is the issue on this thread. If you are grudgingly ok with the approach then let's merge it. Anestis is pressuring to get the Liverpool change on which this depends out to the CKR team.